### DISCUSSION TITLE
Ability to "close" sockets without dropping pending requests

### DISCUSSION THREAD

**Discussion by tayler-king** [uNetworking/uWebSockets.js#874] [Q&A] [answered]
To preface: assume no use of websockets, purely HTTP connections

I'm putting together an application that can handle automatic scaling of uWS server instances based on load. I'm binding each instance to the same port and letting `SO_REUSEPORT` do its thing.

As an example, one method of scaling servers down would be to choose the uWS instance with the least amount of pending requests. By pending request I mean a request that has been accepted by the server and is being processed by the application, i.e. it is non-complete and has not finished responding to the client. Once all pending requests have been completed the instance would terminate.

The issue with the above approach is that I cannot fathom of a way to cleanly handling blocking of new requests when a uWS instance has been marked to be terminated, whilst using `SO_REUSEPORT`.

Would the better option be to just run each server on its own port and have the load balancer handle failing over if a request is rejected with status 503 for example? I would like to avoid this as then I would have to deal with the dynamic port selection of servers and assigning them to the load balancer.

**Comment by e3dio** [comment, marked as answer]
> I cannot fathom of a way to cleanly handling blocking of new requests
```javascript
import { App, us_listen_socket_close } from 'uWebSockets.js';

let listenSocket;
const app = App().listen(port, s => listenSocket = s);

// block new requests
us_listen_socket_close(listenSocket);
```

**Reply by tayler-king** [reply]
Will this keep existing connections open?

**Reply by e3dio** [reply]
Yep

**Reply by tayler-king** [reply]
Hah, that was a massive oversight on my behalf then. Thanks!

**Comment by uNetworkingAB** [comment]
We could possibly consider something similar to maxLifetime where you can specify the maximum time a client can hold a http connection. We have it for websockets for this very reason but not for http

**Comment by e3dio** [comment]
@tayler-king Browsers keep connections open with `Connection: Keep Alive` with 10 second server-side [HTTP_IDLE_TIMEOUT](https://github.com/uNetworking/uWebSockets/blob/master/src/HttpContext.h#L43), `us_listen_socket_close` stops new browser connections but still allows new requests/data from same connection which resets the timeout, you can add code to your request/onData handlers to stop new/existing requests as needed when attempting shutdown, `res.close()` and `res.end(body, true)` force close connection, `app.close()` force closes all connections

**Reply by tayler-king** [reply]
Many thanks for the info :)
